manager/integration/tests/test_scheduling.py

```diff
@@ -694,6 +694,91 @@ def test_replica_auto_balance_node_best_effort(client, volume_name):  # NOQA
     check_volume_data(volume, data)
 
 
+def test_replica_auto_balance_with_data_locality(client, volume_name):  # NOQA
+    """
+    Scenario: replica auto-balance should not cause rebuild loop.
+              - replica auto-balance set to `best-effort`
+              - volume data locality set to `best-effort`
+              - volume has 1 replica
+
+    Issue: https://github.com/longhorn/longhorn/issues/4761
+
+    Given no existing volume in the cluster.
+    And set `replica-auto-balance` to `best-effort`.
+    And create a volume:
+        - set data locality to `best-effort`
+        - 1 replica
+
+    When attach the volume to self-node.
+    And wait for the volume to be healthy.
+    Then the only volume replica should be already on the self-node or
+         get rebuilt one time onto the self-node.
+    And volume have 1 replica only and it should be on the self-node.
+         - check 15 times with 1 second wait interval
+
+    When repeat the test for 10 times.
+    Then should pass.
+    """
+    for i in range(10):
+        replica_auto_balance_with_data_locality_test(
+            client, f'{volume_name}-{i}'
+        )
+
+
+def replica_auto_balance_with_data_locality_test(client, volume_name):  # NOQA
+    common.cleanup_all_volumes(client)
+
+    common.update_setting(client,
+                          SETTING_REPLICA_AUTO_BALANCE, "best-effort")
+
+    self_node = get_self_host_id()
+    number_of_replicas = 1
+    volume = client.create_volume(name=volume_name,
+                                  size=str(200 * Mi),
+                                  numberOfReplicas=number_of_replicas,
+                                  dataLocality="best-effort")
+    volume = common.wait_for_volume_detached(client, volume_name)
+
+    volume.attach(hostId=self_node)
+    volume = wait_for_volume_healthy(client, volume_name)
+
+    for _ in range(30):
+        volume = client.by_id_volume(volume_name)
+        if len(volume.replicas) == number_of_replicas and \
+                volume.replicas[0]['hostId'] == self_node:
+            break
+        try:
+            if len(volume.replicas) == number_of_replicas + 1:
+                is_rebuilded = True
+            assert len(volume.replicas) == number_of_replicas
```

Comment:
I think this assert can make the test flaky sometimes. If the replica on the self node was just built and the replica from another node is not yet removed?
Do you have result from our CI `longhorn-tests-regression`?

Reply:
> I think this assert can make the test flaky sometimes. If the replica on the self node was just built and the replica from another node is not yet removed? 

The `except AssertionError` should catch the assert and retry in the for loop statement.

Do you have result from our CI `longhorn-tests-regression`?

https://github.com/longhorn/longhorn-manager/pull/1526#issuecomment-1286686066